### ISSUE TITLE
Proposal: SafeStaticSingleton

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#4286]
We were running into issues with static objects being destroyed after exiting main. We have the push_finalize_hook to register stuff with Kokkos finalize. What we could do is provide a wrapper, which makes sure destructors of these static objects are called upon finalizing Kokkos. 

```
template<class T>
struct SafeStaticSingleton {
  static T& get() {
    T* ptr = nullptr;
    if(ptr == nullptr) {
       ptr = new T();
       Kokkos::push_finalize_hook([=] () { delete ptr; });
    }
    return *ptr;
  }
};
```

Now if you have some file foo.cpp:
```
class Bar {
   Kokkos::View<double*> b;
   void do_something() { b.resize(31910) }
};

static Kokkos::SafeStaticSingleton<Bar> bar;

void foo() {
   bar.get().do_something();
}
```

This will now deallocate Bar::b in time before Kokkos::finalize gets called.

**Comment by dalg24** [comment]
You might want to declare `ptr` as static.

**Comment by dalg24** [comment]
Please clarify.  Are you proposing a user facing or an internal facility?

**Comment by masterleinad** [comment]
I think it's a good idea to provide something like this but I think you want
``` C++
template<class T>
struct SafeStaticSingleton {
  static T& get() {
    static T* ptr = [](){
      T* new_ptr = new T();
      Kokkos::push_finalize_hook([=] () { delete new_ptr; });
      return new_ptr;
    } ();
    return *ptr;
  }
};
```
so you don't call reallocate memory every time `get()` is called, also see https://godbolt.org/z/bjzb1joTc.

**Comment by ndellingwood** [comment]
Possibly relevant: #4294

**Comment by DavidPoliakoff** [comment]
Commenting here in addition to Slack: I think "get" should check if Kokkos is finalized and error out if it is. Possibly also accept a name on construction, and on get-after-finalize print "error: access to [name] after Kokkos finalization."

**Comment by crtrott** [comment]
I am proposing a public user interface. @masterleinad I don't get what you mean with "so you don't call reallocate memory every time". It only allocates if ptr == nullptr, which should only be the case in the first call or?

**Comment by crtrott** [comment]
I kinda like the idea of adding a name, but am not sure about the interface for that.

**Comment by masterleinad** [comment]
> I am proposing a public user interface. @masterleinad I don't get what you mean with "so you don't call reallocate memory every time". It only allocates if ptr == nullptr, which should only be the case in the first call or?

In your code example you have
```
T* ptr = nullptr;
```
instead of
```
static T* ptr = nullptr;
```
so you always create a local variable that is initialized with a `nullptr` and thus allocates memory in every function call.

**Comment by DavidPoliakoff** [comment]
> I kinda like the idea of adding a name, but am not sure about the interface for that.

Adapting Daniel's sample because it exists: https://godbolt.org/z/K4ahfK3aG . Requires you to make a struct. But I think it would be valuable in debugging, and doesn't impose a cost if you don't want to use it, you just get a shittier debugging experience... with RTTI, we could make the default namer actually use the typeinfo. Options exist

**Comment by ajpowelsnl** [comment]
Morning -- @DavidPoliakoff , @dalg24 , @masterleinad , @crtrott , @ndellingwood  -- I'll add it to our items to discuss on Wednesday.

**Comment by dalg24** [comment]
IMO, the real issue is our not advertising `Kokkos::push_finalize_hook()`.  Neither the [programming guide](https://github.com/kokkos/kokkos/wiki/Initialization) nor the [API reference](https://github.com/kokkos/kokkos/wiki/initialize-and-finalize) mention it.  I object our

**Comment by crtrott** [comment]
@masterleinad No I don't ... :-P 

That was a (now corrected) error, I obviously meant ptr to be a static variable :-)

**Comment by masterleinad** [comment]
I think the problem we are trying to solve here is sufficiently difficult tp help users out but I am not quite sure if this functionality should live in the `Kokkos` namespace or somewhere in the Wiki or similar.

**Comment by DavidPoliakoff** [comment]
@masterleinad : yeah. I think things that are similar, we've said "ehhhh, should this be in Kokkos" and decided on "no." I've generally been in favor of having it in Kokkos, but you're right from a consistency perspective

**Comment by mhoemmen** [comment]
@dalg24 Did the wiki actually exist when I came up with `push_finalize_hook`?  In any case, I must have subconsciously assumed that users would look for the analog of `atexit` and find it, but perhaps it was too bold for me to assume that users know about `atexit`.

**Comment by crtrott** [comment]
If we add this should we add a special namespace?
SF F N A SA
6 4 6 0 0

Should we add this to the code in a special namespace ?
SF F N A SA
0 4 3 4 2

No consensus to add this. We just gonna document a use case.

**Comment by ajpowelsnl** [comment]
@crtrott -- Should we assign this to someone (i.e., the documenting of a use case)?
